scvi/model/base/_rnamixin.py

```diff
@@ -122,34 +225,63 @@ def get_normalized_expression(
             generative_output_key = "scale"
             scaling = library_size
 
+        store_distributions = weights == "importance"
+        if store_distributions and len(transform_batch) > 1:
+            raise NotImplementedError(
+                "Importance weights cannot be computed when expression levels are averaged across batches."
+            )
+
         exprs = []
+        zs = []
+        dist_store = DistributionsConcatenator()
         for tensors in scdl:
             per_batch_exprs = []
             for batch in transform_batch:
                 generative_kwargs = self._get_transform_batch_gen_kwargs(batch)
                 inference_kwargs = {"n_samples": n_samples}
-                _, generative_outputs = self.module.forward(
+                inference_outputs, generative_outputs = self.module.forward(
                     tensors=tensors,
                     inference_kwargs=inference_kwargs,
                     generative_kwargs=generative_kwargs,
                     compute_loss=False,
                 )
-                output = getattr(generative_outputs["px"], generative_output_key)
-                output = output[..., gene_mask]
-                output *= scaling
-                output = output.cpu().numpy()
-                per_batch_exprs.append(output)
-            per_batch_exprs = np.stack(
-                per_batch_exprs
-            )  # shape is (len(transform_batch) x batch_size x n_var)
-            exprs += [per_batch_exprs.mean(0)]
-
-        if n_samples > 1:
-            # The -2 axis correspond to cells.
-            exprs = np.concatenate(exprs, axis=-2)
-        else:
-            exprs = np.concatenate(exprs, axis=0)
-        if n_samples > 1 and return_mean:
+                exp_ = getattr(generative_outputs["px"], generative_output_key)
+                exp_ = exp_[..., gene_mask]
+                exp_ *= scaling
+                per_batch_exprs.append(exp_[None].cpu())
+                if store_distributions:
+                    dist_store.add_distributions(
+                        {
+                            **inference_outputs,
+                            **generative_outputs,
+                        }
+                    )
```

Comment:
This can be more clear. I don't like so much the filtering of distributions inside this object as we already know which two we want